### ISSUE TITLE
Update errors not reported by portmaster-start

### ISSUE THREAD

**Issue by ppacher** [safing/portmaster#115]
**Checklist**:

- [x] I'm using the official portmaster release (i.e no custom builds)
- [x] I searched for similar/existing issues first.
- [x] My issue is not mentioned in the Known Issues section of my [OS](https://github.com/safing/portmaster/wiki)

**What happened**:

`/var/lib/portmaster/portmaster-start --data /var/lib/portmaster update` did not report any errors although it failed to correctly download the required portmaster modules.

**What you expected to happen**:

The update to either succeed or log an error message. (If failed because of missing permissions, i.e. no `sudo`).

**How to reproduce it (as minimally and precisely as possible)**:

1. Download latest installer from safing.io
2. Stop networking (downloading during install should fail) or `export PM_SKIP_DOWNLOAD=yes`
3. Install .dep via dpkg (there should be a warning)
4. `/var/lib/portmaster/updates` **should not** exist
5. Run the update manually: `/var/lib/portmaster/portmaster-start --data /var/lib/portmaster update`
6. `/var/lib/portmaster/udpates` **should** exist but does not. (and no error was logged during 5).

**Environment**:

Portmaster Version: 

<details>
  <summary>portmaster-start version</summary>

```
Portmaster Start
version 0.4.1

commit tags/v0.4.17-0-g04c846a0da3df4fc3fda02afdc366c102f447c54
built with go1.14.2 (gc) linux/amd64
  using options 
  by user@docker
  on 24.07.2020

Licensed under the AGPLv3 license.
The source code is available here: https://github.com/safing/portmaster
```
</details>

Operating System:
- [ ] Windows 7
- [ ] Windows 8/8.1
- [ ] Windows 10
- [x] Linux
  - latest Ubuntu 20.04

**Comment by northys** [comment]
@ppacher 

ad 4) it does. I didn't delete `/var/lib/portmaster` during "CLEANUP" mentioned in origin issue

```
northys at northys-laptop in /v/l/p/updates
» pwd
/var/lib/portmaster/updates
northys at northys-laptop in /v/l/p/updates
» ll
total 16K
drwxr-xr-x 4 root root 4,0K srp  3 00:08 all/
drwxr-xr-x 6 root root 4,0K srp  3 00:08 linux_amd64/
-rw-r--r-- 1 root root 1,7K srp  3 00:08 stable.json
drwx------ 2 root root 4,0K srp  3 00:08 tmp/
```

OS Info:
```
northys at northys-laptop in /v/l/p/updates
» cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
northys at northys-laptop in /v/l/p/updates
» uname -a
Linux northys-laptop 5.4.0-42-generic #46-Ubuntu SMP Fri Jul 10 00:24:02 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by ppacher** [comment]
Thank for the update.

Seems like hard errors are correctly reported. Though, there's an issue with reporting warnings so I'll create a PR for that.

**Comment by northys** [comment]
@ppacher should I try to get portmaster working again by reinstalling with the latest deb from safing.io?

**Comment by ppacher** [comment]
Hi @northys,
thanks for watching :+1: 

It would be great if you could test the new installer form safing.io.

If you still have the version from the first install (0.0.0-1) installed then the new one will warn and refuse to install as the old version is broken and must be purged first. So to be on the safe side, give it an `apt purge portmaster` beforehand and reboot.

The installer also updates `portmaster.service` (again) so if you stop portmaster it will bring systemd-resolved back up so DNS should continue to work (at least on systems using systemd-resolved).

**Comment by northys** [comment]
Hi @ppacher,

it works. I have some questions regarding the upgrade:

1) the safing update command do not output any information. it would be great to se something like "downloaded module xyz"
2) the internet was not working like a minute after installation. would it be possible for dpkg to wait until the upgrade process finishes?

```
● portmaster.service - Portmaster by Safing
     Loaded: loaded (/lib/systemd/system/portmaster.service; enabled; vendor preset: enabled)
     Active: active (running) since Wed 2020-08-05 01:38:39 CEST; 31s ago
       Docs: https://safing.io
             https://docs.safing.io
   Main PID: 65533 (portmaster-star)
      Tasks: 32 (limit: 18956)
     Memory: 368.1M
     CGroup: /system.slice/portmaster.service
             ├─65533 /var/lib/portmaster/portmaster-start --data /var/lib/portmaster core -- --log info
             └─65558 /var/lib/portmaster/updates/linux_amd64/core/portmaster-core_v0-4-18 --data /var/lib/portmaster --log info

srp 05 01:39:04 northys-laptop portmaster-start[65533]: 200805 01:39:04.950 connection:207 ▶ INFO 067 filter: granting connection Unknown::-1 to one.one.one.one. (<nil>), endpoint is not blacklisted (default=permit)
srp 05 01:39:04 northys-laptop portmaster-start[65533]: 200805 01:39:04.950 sts/lookup:021 ▶ WARN 068 intel/filterlists: not searching for cache:intel/filterlists/lists/domain/one.one.one.one. because filterlists not loaded
srp 05 01:39:04 northys-laptop portmaster-start[65533]: 200805 01:39:04.950 sts/lookup:021 ▶ WARN 069 intel/filterlists: not searching for cache:intel/filterlists/lists/domain/one.one.one. because filterlists not loaded
srp 05 01:39:04 northys-laptop portmaster-start[65533]: 200805 01:39:04.950 sts/lookup:021 ▶ WARN 070 intel/filterlists: not searching for cache:intel/filterlists/lists/domain/one.one. because filterlists not loaded
srp 05 01:39:04 northys-laptop portmaster-start[65533]: 200805 01:39:04.950 connection:207 ▶ INFO 071 filter: granting connection Unknown::-1 to one.one.one.one. (<nil>), endpoint is not blacklisted (default=permit)
srp 05 01:39:05 northys-laptop portmaster-start[65533]: 200805 01:39:05.039 tel/entity:170 ▶ WARN 072 intel: failed to resolve IP 95.216.140.122: failed to resolve 122.140.216.95.in-addr.arpa.PTR: all 2 query-compliant resolvers failed, last error: read udp 192.168.31.220:56549->192.168.31.1:53: i/o timeout
srp 05 01:39:05 northys-laptop portmaster-start[65533]: 200805 01:39:05.040 tel/entity:170 ▶ WARN 073 intel: failed to resolve IP 95.216.140.122: failed to resolve 122.140.216.95.in-addr.arpa.PTR: all 2 query-compliant resolvers failed, last error: read udp 192.168.31.220:43144->192.168.31.1:53: i/o timeout
srp 05 01:39:05 northys-laptop portmaster-start[65533]: 200805 01:39:05.040 tel/entity:170 ▶ WARN 074 intel: failed to resolve IP 95.216.140.122: failed to resolve 122.140.216.95.in-addr.arpa.PTR: all 2 query-compliant resolvers failed, last error: read udp 192.168.31.220:60040->192.168.31.1:53: i/o timeout
srp 05 01:39:08 northys-laptop portmaster-start[65533]: 200805 01:39:08.951 connection:263 ▶ INFO 075 filter: dropping connection Unknown::-1 to one.one.one.one. (<nil>) because of an internal error: failed to resolve: all 2 query-compliant resolvers failed, last error: read udp 192.168.31.220:47597->192.168.31.1:53: i/o timeout
srp 05 01:39:08 northys-laptop portmaster-start[65533]: 200805 01:39:08.951 connection:263 ▶ INFO 076 filter: dropping connection Unknown::-1 to one.one.one.one. (<nil>) because of an internal error: failed to resolve: all 2 query-compliant resolvers failed, last error: read udp 192.168.31.220:12811->192.168.31.1:53: i/o timeout
```

When dashboard opened after ~30s there was note that networking may be slow during upgrade. See components box.

![Screenshot from 2020-08-05 01-39-29](https://user-images.githubusercontent.com/2878126/89356068-2966fe00-d6bd-11ea-9ae8-372225908eee.png)

**Comment by ppacher** [comment]
Hi,

I'm glad it worked for you this time @northys!

Let's shed some light on your questions:

> 1. the safing update command do not output any information. it would be great to see something like "downloaded module xyz"

The PR that closed this issue (#117) adds exactly that behavior (actually it was always there but we missed to flush log messages to the screen). As we released the .deb package earlier yesterday it did not yet include the new `portmaster-start` binary with this patch applied. Though, since then the Portmaster service should have already picked up the new version and downloaded it. 

![update](https://user-images.githubusercontent.com/1043756/89378608-75816500-d6f4-11ea-8317-7d4bbef3f9ce.png)

> 2. ~~the internet was not working like a minute after installation. would it be possible for `dpkg` to wait until the upgrade process finishes?~~ this is nonsence. portmaster is firewall and firewall should not pass anything during the allowlist update / not available.

We also though about how to handle requests that appear while allow/blocklists are being updated and processed. The problem is that updating the filterlists requires Portmaster to build up a lookup cache to speed up the decision making. However, building this lookup caches is (at least for the very first time), time consuming so we actually decided to **not block** requests that happen during the very first filterlist upgrade/initialization. You connection should actually have worked just `dpkg` finished. From the log messages, it seems like you upstream resolver at 192.168.31.1 timed out which could be a strange coincidence or point to some timing issues inside the Portmaster.

I'll investigate that, thanks for your report :+1:
